File: src/main/java/com/tobe/healthy/push/domain/entity/MemberToken.java
package com.tobe.healthy.push.domain.entity;

import com.tobe.healthy.common.BaseTimeEntity;
import com.tobe.healthy.member.domain.entity.Member;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class MemberToken extends BaseTimeEntity<MemberToken, Long> {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "member_token_id")
    private Long id;

    @Enumerated(EnumType.STRING)
    private DeviceType deviceType;

    private String token;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id")
    @ToString.Exclude
    private Member member;

    private MemberToken(Member member, String token, DeviceType deviceType) {
        this.member = member;
        this.token = token;
        this.deviceType = deviceType;
    }

    public void changeToken(String token, DeviceType deviceType) {
        this.token = token;
        this.deviceType = deviceType;
    }

    public static MemberToken register(Member member, String token, DeviceType deviceType) {
        return new MemberToken(member, token, deviceType);
    }
}

File: src/main/kotlin/com/tobe/healthy/push/domain/entity/MemberToken.kt
package com.tobe.healthy.push.domain.entity

import com.tobe.healthy.common.BaseTimeEntity
import com.tobe.healthy.member.domain.entity.Member
import jakarta.persistence.*
import jakarta.persistence.EnumType.STRING
import jakarta.persistence.FetchType.LAZY
import jakarta.persistence.GenerationType.IDENTITY
import lombok.ToString

@Entity
class MemberToken(

    @Enumerated(STRING)
    var deviceType: DeviceType,

    var token: String,

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "member_id")
    @ToString.Exclude
    val member: Member? = null,

    @Id
    @GeneratedValue(strategy = IDENTITY)
    @Column(name = "member_token_id")
    val id: Long? = null

) : BaseTimeEntity<MemberToken, Long>() {

    fun changeToken(token: String, deviceType: DeviceType) {
        this.token = token
        this.deviceType = deviceType
    }

    companion object {
        fun register(member: Member, token: String, deviceType: DeviceType): MemberToken {
            return MemberToken(
                member = member,
                token = token,
                deviceType = deviceType
            )
        }
    }
}